rocketpy/Function.py

```diff
@@ -2030,3 +2118,101 @@ def differentiate(self, x, dx=1e-6):
         # h = (10)**-300
         # z = x + h*1j
         # return self(z).imag/h
+
+
+def funcify_method(*args, **kwargs):
+    """Decorator factory to wrap methods as Function objects and save them as cached
+    properties.
+
+    Parameters
+    ----------
+    *args : list
+        Positional arguments to be passed to rocketpy.Function.
+    **kwargs : dict
+        Keyword arguments to be passed to rocketpy.Function.
+
+    Returns
+    -------
+    decorator : function
+        Decorator function to wrap callables as Function objects.
+
+    Examples
+    --------
+    >>> class Test():
+    ...     @Funcify(inputs=['x'], outputs=['y'])
+    ...     def func(x):
+    ...         return x**2
+    >>> t = Test()
+    ... t.func
+    Function from R1 to R1 : (x) → (y)
+    >>> g = 2*t.func + 3
+    >>> g(2)
+    11
+
+    Can also be used without any arguments:
+
+    >>> class Test():
+    ...     @Funcify
+    ...     def func(x):
+    ...         return x**2
+    >>> t = Test()
+    ... t.func
+    Function from R1 to R1 : (Scalar) → (Scalar)
+
+    Can also be used when the method already returns a Function instance. In such case
+    it is interesting to use the `inputs` and `outputs` arguments to overwrite the
+    inputs and outputs of the method.
+
+    >>> class Test():
+    ...     @Funcify(inputs='x', outputs='y')
+    ...     def func(x):
+    ...         return Function(lambda x: x**2)
+    >>> t = Test()
+    ... t.func
+    Function from R1 to R1 : (x) → (y)
+
+    In order to reset the cache, just delete de attribute from the instance:
+
+    >>> del t.func
+
+    Once it is requested again, it will be re-created as a new Function object:
+
+    >>> t.func
+    Function from R1 to R1 : (Scalar) → (Scalar)
+    """
+    func = None
+    if len(args) == 1 and callable(args[0]):
+        func = args[0]
+        args = []
+
+    class funcify_method_decorator:
+        def __init__(self, func):
+            self.func = func
+            self.attrname = None
+            self.__doc__ = func.__doc__
+
+        def __set_name__(self, owner, name):
+            self.attrname = name
+
+        def __get__(self, instance, owner=None):
+            if instance is None:
+                return self
+            cache = instance.__dict__
+            try:
+                val = cache[self.attrname]
+            except KeyError:
+                source = self.func(instance)
```

Comment:
Found a not so elegant fix with the following try clause:

```
try:
    source = self.func(instance)
except TypeError:
    source = lambda *_: self.func(instance, *_)
```

Reply:
Great suggestion! I could not find anything better really, so I restructure the code in a single try except block that handles all possible scenarios. Seems elegant to me.

Reply:
Mark as resolved if you are satisfied.